### PR TITLE
Move completions rendering into a separate module

### PR DESCRIPTION
**Pull request by popzxc**
This PR extracts rendering-related things from `Completions` structure to the new `render` module.

`render` module declares a `Render` structure (which is a generic renderer interface), `RenderContext` (interface for data/methods not required for completions generating, but required for rendering), and a bunch of smaller `*Render` structures which encapsulate logic behind rendering a certain item.

This is just a step in full separation direction, since the following this are still to be done:

- Move some data from `CompletionContext` to the `RenderContext`;
- Forbid any kind of rendering outside of `render` module;
- Extract score computing into a separate module.

This PR is already pretty big, so not to make it even harder to review I decided to split this process into several subsequent PRs.

### CODE REVIEW

File: crates/completion/src/completions.rs [hunk 1]
```diff
@@ -189,50 +78,13 @@ impl Completions {
...
-            }
-            None if needs_bang => builder.insert_text(format!("{}!", name)),
-            _ => {
-                mark::hit!(dont_insert_macro_call_parens_unncessary);
-                builder.insert_text(name)
-            }
-        };
-
-        self.add(builder.build());
+        if let Some(item) = MacroRender::new(ctx.into(), name, macro_).render() {
```
Comment (matklad) [inline crates/completion/src/completions.rs:85]: This is a doer object anti-pattern, see #6437

**Comment by matklad** [comment]
Looks good to me, this is definitely a direction worth exploring. I like the stricter separation, and I like that each thing gets its dedicated module.  

I slightly don't like the proliferation of types here. There's a natural tension between "every bit of logic is separated into a module" and "there's only few vocabulary types". In an ideal world, there should be many pieces, but they all should operate the same types. 

I am also not sure if super-strict separation would be benefitial long-term (although i expect that it would).


The only not-so-minor complaint I have is about do-er objects, the rest looks good to me!

bors d+

File: crates/completion/src/completions.rs (hunk 1 above)
Comment (popzxc) [reply crates/completion/src/completions.rs:85]: Yep, that makes sense.

**Pull request merged**